src/Architecture.cpp

```diff
@@ -44,6 +44,8 @@ void Architecture::loadCouplingMap(std::istream&& is) {
   if (std::getline(is, line)) {
     if (std::regex_search(line, m, rNqubits)) {
       nqubits = static_cast<std::uint16_t>(std::stoul(m.str(1)));
+      zeroMatrix.clear();
+      zeroMatrix.resize(nqubits, std::vector<double>(nqubits, 0.0));
```

Comment:
This solution really is not too elegant. Though my main problem with it is not space concerns, but that `zeroMatrix` needs to resized, wherever `nqubits` is changed/reset and all the maintainability problems that come with that.

My reasoning for this implementation was as follows: 
`fidelityDistanceTables` for a number of edge skips, that is higher than the span of the coupling graph, will always only consist of 0s. Therefore, `fidelityDistanceTables` are only stored up to that point, and after that, the architecture should just return a 0-matrix. One could either generate this 0-matrix on the spot each time `getFidelityDistanceTable` is called with such a high edge skipping number, or we could keep a stand-in 0-matrix around. I went with the latter for efficiency, although realizing now, that I only use `getFidelityDistanceTable` in the tests and not the mapper itself, I tend more to the former solution for maintainability.

Using `std::optional` is a rather unintuitive solution in my opinion, as there should be no real difference (apart from the entry values) between a distance table with 0 edge skips or 100 edge skips. Distance tables are not _undefined_ for high edge skipping numbers, they are just redundant enough to not be explicitly saved behind the scenes, but they are still a `nqubits`x`nqubits` matrix.

Reply:
Thanks for the reasoning behind this. That makes sense.
If that method is really only used in the tests, then I believe I would also prefer thread-hoc construction method for the zero matrix.
Just to play around with a potential idea: These matrices aren't really that sparse, right? Otherwise, one could just use a sparse matrix to represent the distances and at that point, the zero matrix becomes trivially compact.

Reply:
Thanks for the careful analysis. In that case, I would go even further and say that we should definitely stick with the dense representation. 
Quantum architectures are getting bigger and bigger and it will be interesting to apply mappers to 100-1000 qubit devices.